File: java/anagram/src/main/java/Anagram.java
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;

class Anagram {

  String _word;

  Anagram(String s) {
    _word = s.toLowerCase();
  }

  List<String> match(List<String> candidates) {
    List<String> matches = new ArrayList<String>();
    for (String candidate : candidates) {
      if (isAnagram(candidate.toLowerCase())) {
        matches.add(candidate);
      }
    }
    return matches;
  }

  boolean isAnagram(String s) {
    System.out.println("in isAnagram, " + s);
    if (s.length() != _word.length() || s.equals(_word)) {
      return false;
    }
    char[] _wordArray = _word.toCharArray();
    char[] sArray = s.toCharArray();
    Arrays.sort(_wordArray);
    Arrays.sort(sArray);
    return new String(_wordArray).equals(new String(sArray));
  }
}

File: typescript/anagram/anagram.ts
class Anagram {
  _word: string;

  constructor(word: string) {
    this._word = word.toLowerCase();
  }

  isAnagram(candidate: string) {
    if (this._word === candidate) return false;
    return ([...candidate].sort().join('') === [...this._word].sort().join(''));
  }

  matches(...candidates: string[]) {
    return candidates.reduce((accum: string[], candidate) => {
      if (this.isAnagram(candidate.toLowerCase())) {
        accum.push(candidate);
      }
      return accum;
    }, []);
  }
}

export default Anagram;